display_list/geometry/dl_region.cc

```diff
@@ -8,31 +8,444 @@
+    new_capacity = std::max(new_capacity, size_t(512));
+    reserve(new_capacity);
+  }
+  SpanChunkHandle res = size_;
+  size_ += chunk_size + 1;
+  spans_[res].left = chunk_size;
+
+  auto* dst = spans_ + res + 1;
+  memmove(dst, begin, chunk_size * sizeof(Span));
+
+  return res;
+}
+
+size_t DlRegion::SpanBuffer::getChunkSize(SpanChunkHandle handle) const {
+  FML_DCHECK(handle < size_);
+  return spans_[handle].left;
+}
+
+void DlRegion::SpanBuffer::getSpans(SpanChunkHandle handle,
+                                    const DlRegion::Span*& begin,
+                                    const DlRegion::Span*& end) const {
+  FML_DCHECK(handle < size_);
+  auto& info = spans_[handle];
+  begin = spans_ + handle + 1;
+  end = begin + info.left;
+}
+
+DlRegion::DlRegion(const std::vector<SkIRect>& rects) {
+  addRects(rects);
+}
+
+DlRegion::DlRegion() {}
+
+bool DlRegion::spansEqual(SpanLine& line,
+                          const Span* begin,
+                          const Span* end) const {
+  const Span *our_begin, *our_end;
+  span_buffer_.getSpans(line.chunk_handle, our_begin, our_end);
+  size_t our_size = our_end - our_begin;
+  size_t their_size = end - begin;
+  if (our_size != their_size) {
+    return false;
+  }
+
+  return memcmp(our_begin, begin, our_size * sizeof(Span)) == 0;
+}
+
+DlRegion::SpanLine DlRegion::makeLine(int32_t top,
+                                      int32_t bottom,
+                                      const SpanVec& v) {
+  return makeLine(top, bottom, v.data(), v.data() + v.size());
+}
+
+DlRegion::SpanLine DlRegion::makeLine(int32_t top,
+                                      int32_t bottom,
+                                      const Span* begin,
+                                      const Span* end) {
+  auto handle = span_buffer_.storeChunk(begin, end);
+  return {top, bottom, handle};
+}
+
+size_t DlRegion::mergeLines(std::vector<Span>& res,
+                            const SpanBuffer& a_buffer,
+                            SpanChunkHandle a_handle,
+                            const SpanBuffer& b_buffer,
+                            SpanChunkHandle b_handle) {
+  const Span *begin1, *end1;
+  a_buffer.getSpans(a_handle, begin1, end1);
+
+  const Span *begin2, *end2;
+  b_buffer.getSpans(b_handle, begin2, end2);
+
+  size_t min_size = (end1 - begin1) + (end2 - begin2);
+  if (res.size() < min_size) {
+    res.resize(min_size);
+  }
+  Span* end = res.data();
+
+  while (true) {
+    if (begin1->right < begin2->left - 1) {
+      *end = *begin1;
+      ++begin1;
+      ++end;
+      if (begin1 == end1) {
+        break;
+      }
+    } else if (begin2->right < begin1->left) {
+      *end = *begin2;
+      ++begin2;
+      ++end;
+      if (begin2 == end2) {
+        break;
+      }
+    } else {
+      break;
+    }
+  }
+
+  Span currentSpan{0, 0};
+  while (begin1 != end1 && begin2 != end2) {
+    if (currentSpan.left == currentSpan.right) {
+      if (begin1->right < begin2->left - 1) {
```

Comment:
Why `- 1`?

Reply:
I think this just falls through to setting up `currentSpan` instead and merging from there, but that costs a bunch of extra tests to accomplish the same thing.